js/editor.js

```diff
@@ -201,9 +172,47 @@ function set_syntax(editor_id, select_id) {
   });
 }
 
+function json_to_session(editor, state) {
+  var sess = editor.getSession();
+  sess.setValue(state.content);
+  editor.selection.fromJSON(state.selection);
+  sess.setOptions(state.options);
+  sess.setMode(state.mode);
+  sess.setScrollTop(state.scrollTop);
+  sess.setScrollLeft(state.scrollLeft);
+  sess.$undoManager.$undoStack = state.history.undo;
+  sess.$undoManager.$redoStack = state.history.redo;
+}
+
+function session_to_json(editor) {
+  return {
+    content: editor.getSession().getValue(),
+    selection: editor.getSelection().toJSON(),
+    options: editor.getSession().getOptions(),
+    mode: editor.session.getMode().$id,
+    scrollTop: editor.session.getScrollTop(),
+    scrollLeft: editor.session.getScrollLeft(),
+    history: {
+      undo: editor.session.getUndoManager().$undoStack,
+      redo: editor.session.getUndoManager().$redoStack
+    },
+  }
+}
+
+function get_editor_session(editor, init) {
+  var key = editor.container.id;
+  var state = JSON.parse(localStorage.getItem(init ? key + init_name : key));
+  if (!state) return;
+  json_to_session(editor, state);
+}
+
 function reset_localstorage(editor_id) {
   try {
+    var editor = ace.edit(editor_id);
+    sess = JSON.parse(localStorage.getItem(editor_id + init_name));
+    sess.mode = editor.getSession().getMode().$id;
+    localStorage.setItem(editor_id + init_name, JSON.stringify(sess));
```

Comment:
Actually we don't know the mode when set the init state. That's why we grab the last one from current session